# Flow cell stats

```{r setup,warning=F}
library(tidyverse)
```

## Per sample

Now we've QC'd and pooled the samples, may as well use this.

```{r load}
cov = read.delim("data/2017-09-28_flowcell1-stats/nova-fc1-qc-readstats.tsv", stringsAsFactors=F) %>%
    mutate(sample = sub(".fastq.gz", "", basename(filename)),
           coverage = bases / 640e6)
kmers = read.delim("data/2017-09-28_flowcell1-stats/nova-fc1-unique-kmers.tsv", stringsAsFactors=F) %>%
    mutate(sample = sub(".fastq.gz", "", basename(filename)))

md = left_join(cov, kmers, by=c("sample", "filename")) %>%
    select(-filename) %>%
    select(sample, everything())
write.csv(md, "data/common/nova-fc1-stats.csv", row.names=F)
```

```{r covhist}
ggplot(md, aes(coverage)) + 
    geom_histogram() +
    labs(x="Mean Coverage", y="Number of Samples") +
    theme_bw()
```

```{r rvk}
ggplot(md, aes(bases, unique_kmers)) + 
    geom_point() +
    labs(x="Number of Bases", y="Number of Distinct Kmers") +
    theme_bw()
```


## Per library

This code/data is older, and is per library, pre-QC


```{r loadlib}
d = read.csv("data/2017-09-28_flowcell1-stats/flowcell1-stats.csv")
str(d)
```

```{r}
d.merge = d %>%
    group_by(sample) %>%
    summarise(yield.mb = sum(parse_number(yield.mb)),
              read.num = sum(parse_number(reads.pass))) %>%
    ungroup() %>%
    mutate(coverage = yield.mb / 640,
           sample=trimws(sample)) %>%
    filter(sample != "Undetermined")
write.csv(d.merge, "data/common/nova-fc1-stats.csv", row.names = F)

head(d.merge)
hist(d.merge$yield.mb)
hist(d.merge$coverage)
```

Looks pretty good!

```{r}
summary(d.merge$coverage)
table(d.merge$coverage <1)
sum(d.merge$read.num)
sum(d.merge$yield.mb)
```